### ISSUE TITLE
DataFrame.resample(...).mean() generates NaN for all but first partition?

### ISSUE THREAD

**Issue by SebastianRiedel** [dask/dask#3032]
I encountered a behaviour, which seems like a bug to me. Looking at the code sample below, the .mean() after the resample seems to only work on the first partition and generating NaN for the remaining partitions. The same resample().mean() works if applied to the raw pandas DataFrame. 

Minimal example to reproduce:
```
import numpy as np
import pandas as pd
import dask
import dask.dataframe as dd

print 'numpy', np.__version__
print 'pandas', pd.__version__
print 'dask', dask.__version__
print

days = pd.date_range(start='2000-04-28', end='2017-04-28', freq='D')

dictdf = dict(a=np.random.rand(len(days)),
              b=np.random.rand(len(days))+1,
              c=np.random.rand(len(days))+2)

df = pd.DataFrame(dictdf, index=days)
print 'Pandas DataFrame\n', df.head(), '\n\n----------'

dddf = dd.from_pandas(df, npartitions=10)
print 'Dask DataFrame\n', dddf, '\n\n-----------'

resampled_df = dddf.resample('3M').mean().compute()
print 'Dask Resampled DataFrame\n', resampled_df, '\n\n-----------'
```
Output I get:
```
numpy 1.13.3
pandas 0.21.1
dask 0.16.0

Pandas DataFrame
                   a         b         c
2000-04-28  0.684370  1.662188  2.615896
2000-04-29  0.017562  1.959581  2.956839
2000-04-30  0.964705  1.867709  2.060739
2000-05-01  0.274513  1.541813  2.951707
2000-05-02  0.676925  1.816926  2.450844 

----------
Dask DataFrame
Dask DataFrame Structure:
                      a        b        c
npartitions=10                           
2000-04-28      float64  float64  float64
2002-01-09          ...      ...      ...
...                 ...      ...      ...
2015-08-17          ...      ...      ...
2017-04-28          ...      ...      ...
Dask Name: from_pandas, 10 tasks 

-----------
Dask Resampled DataFrame
                   a         b         c
2000-04-30  0.555546  1.829826  2.544492
2000-07-31  0.494684  1.511440  2.509780
2000-10-31  0.535168  1.513747  2.482575
2001-01-31  0.485873  1.510528  2.510984
2001-04-30  0.523038  1.528395  2.474981
2001-07-31  0.438252  1.513022  2.538671
2001-10-31  0.492840  1.498296  2.481057
2002-01-31       NaN       NaN       NaN
2002-04-30       NaN       NaN       NaN
2002-07-31       NaN       NaN       NaN
2002-10-31       NaN       NaN       NaN
2003-01-31       NaN       NaN       NaN
2003-04-30       NaN       NaN       NaN
2003-07-31       NaN       NaN       NaN
2003-10-31       NaN       NaN       NaN
2004-01-31       NaN       NaN       NaN
2004-04-30       NaN       NaN       NaN
2004-07-31       NaN       NaN       NaN
2004-10-31       NaN       NaN       NaN
2005-01-31       NaN       NaN       NaN
2005-04-30       NaN       NaN       NaN
2005-07-31       NaN       NaN       NaN
2005-10-31       NaN       NaN       NaN
2006-01-31       NaN       NaN       NaN
2006-04-30       NaN       NaN       NaN
2006-07-31       NaN       NaN       NaN
2006-10-31       NaN       NaN       NaN
2007-01-31       NaN       NaN       NaN
2007-04-30       NaN       NaN       NaN
2007-07-31       NaN       NaN       NaN
...              ...       ...       ...
2010-01-31       NaN       NaN       NaN
2010-04-30       NaN       NaN       NaN
2010-07-31       NaN       NaN       NaN
2010-10-31       NaN       NaN       NaN
2011-01-31       NaN       NaN       NaN
2011-04-30       NaN       NaN       NaN
2011-07-31       NaN       NaN       NaN
2011-10-31       NaN       NaN       NaN
2012-01-31       NaN       NaN       NaN
2012-04-30       NaN       NaN       NaN
2012-07-31       NaN       NaN       NaN
2012-10-31       NaN       NaN       NaN
2013-01-31       NaN       NaN       NaN
2013-04-30       NaN       NaN       NaN
2013-07-31       NaN       NaN       NaN
2013-10-31       NaN       NaN       NaN
2014-01-31       NaN       NaN       NaN
2014-04-30       NaN       NaN       NaN
2014-07-31       NaN       NaN       NaN
2014-10-31       NaN       NaN       NaN
2015-01-31       NaN       NaN       NaN
2015-04-30       NaN       NaN       NaN
2015-07-31       NaN       NaN       NaN
2015-10-31       NaN       NaN       NaN
2016-01-31       NaN       NaN       NaN
2016-04-30       NaN       NaN       NaN
2016-07-31       NaN       NaN       NaN
2016-10-31       NaN       NaN       NaN
2017-01-31       NaN       NaN       NaN
2017-04-30       NaN       NaN       NaN

[69 rows x 3 columns] 

-----------
```

**Comment by jcrist** [comment]
Thanks for the excellent bug report. This is indeed a bug in resample. Looks like it fails because `3M` is a non-anchored frequency of > 1 day (resample with `M` works fine because it is an anchored frequency). In the short term we could add a check for this to throw a `NotImplementedError`, but in the long term this should be fixable. It's been sufficiently long since I touched this code, but here's what I remember when I first implemented it:

The intended flow of this method is:
- Determine new divisions to repartition the dataframe to such that no resampled bin spans more than one divisions (`newdivs` in the code)
- Determine divisions that will result after calling `resample` on each partition (`outdivs` in the code). These can differ from `newdivs` due to `label`/`closed` behavior in resample.
- Call `resample` on each partition
- `reindex` each partition to set `nan` for any missing labels (needed to handle missing rows that fall between partitions)

The tricky bit here seems to be that for non-anchored frequencies, the initial row in the partition affects what the output labels are (starting month determines all subsequent indices). This is further complicated because the behavior also differs based on the frequency unit (seconds, days, months, etc...).

Ping @TomAugspurger who might know of a cleaner way of handling this.
